src/status_im/contexts/keycard/events.cljs

```diff
@@ -69,7 +65,53 @@
          {:effects.utils/show-popup {:title "wrong-keycard"}}
          {:db (-> db
                   (assoc-in [:keycard :application-info :pin-retry-counter] pin-retries-count)
-                  (update-in [:keycard :pin] assoc :status :error))
-          :fx [[:dispatch [:keycard/hide-connection-sheet]]
+                  (assoc-in [:keycard :pin :status] :error))
+          :fx [[:dispatch [:keycard/disconnect]]
                (when (zero? pin-retries-count)
                  [:effects.utils/show-popup {:title "frozen-keycard"}])]})))))
+
+(rf/reg-event-fx :keycard/on-get-application-info-success
+ (fn [{:keys [db]} [application-info {:keys [key-uid on-success-fx]}]]
+   (let [error (keycard.utils/validate-application-info key-uid application-info)]
+     (if error
+       (case error
+         :keycard/error.not-keycard
+         {:fx [[:dispatch [:keycard/disconnect]]
+               [:dispatch [:open-modal :screen/keycard.not-keycard]]]}
+         :keycard/error.keycard-blank
+         {:fx [[:dispatch [:keycard/disconnect]]
+               [:dispatch [:open-modal :screen/keycard.empty]]]}
+         {:db (assoc-in db [:keycard :application-info-error] error)
+          :fx [[:dispatch [:keycard/disconnect]]
+               [:dispatch [:open-modal :screen/keycard.error]]]})
+       {:db (-> db
+                (assoc-in [:keycard :application-info] application-info)
+                (assoc-in [:keycard :pin :status] :verifying))
+        :fx on-success-fx}))))
+
+(rf/reg-event-fx :keycard/get-application-info
+ (fn [_ [{:keys [on-success on-failure]}]]
+   {:effects.keycard/get-application-info {:on-success on-success :on-failure on-failure}}))
+
+(rf/reg-event-fx :keycard/cancel-connection
+ (fn [{:keys [db]}]
+   {:db (-> db
+            (assoc-in [:keycard :on-card-connected-event-vector] nil)
+            (assoc-in [:keycard :on-nfc-cancelled-event-vector] nil))}))
+
+(rf/reg-event-fx :keycard/disconnect
+ (fn [_ _]
+   {:fx [[:dispatch [:keycard/cancel-connection]]
+         [:dispatch [:keycard/hide-connection-sheet]]]}))
+
+(rf/reg-event-fx :keycard/connect
+ (fn [{:keys [db]} [args]]
+   (let [connected?   (get-in db [:keycard :card-connected?])
+         event-vector [:keycard/get-application-info
+                       {:on-success #(rf/dispatch [:keycard/on-get-application-info-success % args])}]]
```

Comment:
I think we might need to be careful with storing functions in the database. Maybe it's not a problem in this case, but if we try to serialise the db-state the function we may not be able to deserialise the function later.

Reply:
i would think about this much ) we should do it how it's better now, but I agree, I'm also no a fun of such approach, we could think how we could improve it later

Reply:
later on if we update the keycard effects to use `call-continuation`, then we may be able to do something like this:
```clojure
[:keycard/get-application-info
  {:on-success [:keycard/on-get-application-info-success args]}]
```
But we would need to refactor `:keycard/on-get-application-info-success` to the it's arguments in a different order (I think).